packages/near-membrane-base/src/environment.ts

```diff
@@ -1,192 +1,181 @@
-        // the proper errors without leaking instances into a sandbox
-        const redEnvFactory = redGlobalThis.eval(`(${serializedRedEnvSourceText})`);
-        // Important Note: removing the indirection for apply and construct breaks
-        // chrome karma tests for some unknown reasons. What is seems harmless turns out
-        // to be fatal, why? it seems that this is because Chrome does identity checks
-        // on those intrinsics, and fails if the detached iframe is calling an intrinsic
-        // from another realm.
-        const blueHooks: MarshalHooks = {
-            apply(
-                target: BlueFunction,
-                thisArgument: BlueValue,
-                argumentsList: ArrayLike<BlueValue>
-            ): BlueValue {
-                return ReflectApply(target, thisArgument, argumentsList);
-            },
-            construct(
-                target: BlueConstructor,
-                argumentsList: ArrayLike<BlueValue>,
-                newTarget?: any
-                // eslint-disable-next-line class-methods-use-this
-            ): BlueValue {
-                return ReflectConstruct(target, argumentsList, newTarget);
-            },
+        const { blueConnector: localInit, redConnector: foreignInit, distortionCallback } = options;
+
+        let blueHooks: Parameters<HooksCallback>;
+        let redHooks: Parameters<HooksCallback>;
+
+        const blueExportsCallback: HooksCallback = (...hooks) => {
+            blueHooks = hooks;
+        };
+        const redExportsCallback: HooksCallback = (...hooks) => {
+            redHooks = hooks;
         };
-        this.getRedValue = redEnvFactory(this, blueHooks);
-        this.getBlueValue = blueProxyFactory(this);
-    }
 
-    // istanbul ignore next
-    // eslint-disable-next-line class-methods-use-this
-    getBlueValue(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        red: RedValue
-    ): BlueValue {
-        // placeholder since this will be assigned in construction
-    }
+        const localConnect = localInit(
+            undefinedSymbol,
+            'blue',
+            false,
+            blueExportsCallback,
+            distortionCallback
+        );
+        const foreignConnect = foreignInit(undefinedSymbol, 'red', true, redExportsCallback);
+        localConnect(...redHooks!);
+        foreignConnect(...blueHooks!);
 
-    // istanbul ignore next
-    // eslint-disable-next-line class-methods-use-this
-    getRedValue(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        blue: BlueValue
-    ): RedValue {
-        // placeholder since this will be assigned in construction
+        const [blueGetSelectedTarget, blueGetTransferableValue] = blueHooks!;
+        this.blueGetSelectedTarget = blueGetSelectedTarget;
+        this.blueGetTransferableValue = blueGetTransferableValue;
+        // prettier-ignore
+        const [,,
```

Comment:
Can we add comments to the holes in the array to document what they are.

Reply:
the problem is prettier, which is very rigid, we will see what I can do.

Reply:
I have a better way to do this, which doesn't use "holes" and is much faster :D

Reply:
curious about it.... I did what @jdalton was suggesting, but it is ugly as well because of prettier, so anything better, let me know.